### PR TITLE
[mono] Mono support for InlineArrayAttribute

### PR DESCRIPTION
**Pull request by kotlarmilos**
This PR Implements support for `InlineArrayAtribute` as outlined in https://github.com/dotnet/runtime/issues/61135 and tracked in https://github.com/dotnet/runtime/issues/80798. 

It implements structure layout, updates GC class bitmap to consider the struct field as an array, disallows marshalling structs with the attribute in `JS Interop generator` and adds a corresponding test.

### CODE REVIEW

File: src/mono/mono/metadata/class-init.c [hunk 1]
```diff
@@ -2416,6 +2459,9 @@ mono_class_layout_fields (MonoClass *klass, int base_instance_size, int packing_
 		class_size = field_offsets [i] + size;
 	}
 
+	if (m_class_is_inlinearray (klass) && has_static_fields)
+		mono_class_set_type_load_failure (klass, "Struct with inline array atribute must have a single field.");
```
Comment (lambdageek) [inline src/mono/mono/metadata/class-init.c:2463]: Will we get a TLE for something like

```csharp
[InlineArray(10)]
struct EmptyStruct
{
}
```

Also, is there a reason why we're not allowing `static` fields in an InlineArray struct?  That seems unrelated.  Static fields don't really affect the layout at all, and I could imagine they could be useful (for example to cache some delegates or something).

File: src/mono/mono/metadata/class-init.c [hunk 2]
```diff
@@ -746,65 +766,75 @@ mono_generic_class_setup_parent (MonoClass *klass, MonoClass *gtd)
...
+			MonoMethod *ctor = mono_get_method_checked (image, method_token, NULL, NULL, &error);
+			if (ctor) {
+				const char *data = mono_metadata_blob_heap (image, cols [MONO_CUSTOM_ATTR_VALUE]);
+				uint32_t data_size = mono_metadata_decode_value (data, &data);
+				MonoDecodeCustomAttr *decoded_attr = mono_reflection_create_custom_attr_data_args_noalloc (image, ctor, (guchar*)data, data_size, &error);
+				mono_error_assert_ok (&error);
+				attr->value = decoded_attr->typed_args[0]->value.primitive;
+			} else {
+				g_warning ("Can't find custom attr constructor image: %s mtoken: 0x%08x due to: %s", image->name, method_token, mono_error_get_message (&error));
+			}
```
Comment (lambdageek) [inline src/mono/mono/metadata/class-init.c:786]: I would prefer to pull this out into a callback for InlineArrayAttribute specifically.

I think this also needs some validation: the ctor has 1 argument, the argument really is a primitive, there are no named properties, stuff like that.
otherwise this might be leaking decoded_attr fields.

Actually do you need to free `decoded_attr`?

**Review by lambdageek — COMMENTED** [review]
Most of the structure is looking pretty reasonable

File: src/mono/mono/metadata/object.c [hunk 3]
```diff
@@ -878,6 +878,11 @@ compute_class_bitmap (MonoClass *klass, gsize *bitmap, int size, int offset, int
 			pos += offset;
 
 			type = mono_type_get_underlying_type (field->type);
+
+			// If a struct has inline array attribute and type is not generic instance, consider it as an array
+			if (m_class_is_inlinearray (klass) && type->type != MONO_TYPE_GENERICINST && type->type != MONO_TYPE_TYPEDBYREF && type->type != MONO_TYPE_VALUETYPE)
+				type->type = MONO_TYPE_ARRAY;
```
Comment (lambdageek) [inline src/mono/mono/metadata/object.c:884]: 1. I don't think it's a great idea to modify type->type.  I think you're actually altering `field->type->type`.  That will probably break something unrelated.

2. Why are generic instances different?

File: src/mono/mono/metadata/class-private-definition.h [hunk 4]
```diff
@@ -49,6 +49,8 @@ struct _MonoClass {
 	guint wastypebuilder  : 1; /* class was created at runtime from a TypeBuilder */
 	guint is_array_special_interface : 1; /* gtd or ginst of once of the magic interfaces that arrays implement */
 	guint is_byreflike    : 1; /* class is a valuetype and has System.Runtime.CompilerServices.IsByRefLikeAttribute */
+	guint is_inlinearray    : 1; /* class is a valuetype and has System.Runtime.CompilerServices.InlineArrayAttribute */
+	guint inlinearray_value; /* class is a valuetype and has System.Runtime.CompilerServices.InlineArrayAttribute value */
```
Comment (vargaz) [inline src/mono/mono/metadata/class-private-definition.h:53]: This should be after the bitflags.

File: src/mono/mono/metadata/class-getters.h [hunk 5]
```diff
@@ -29,6 +29,8 @@ MONO_CLASS_GETTER(m_class_is_unicode, gboolean, , MonoClass, unicode)
 MONO_CLASS_GETTER(m_class_was_typebuilder, gboolean, , MonoClass, wastypebuilder)
 MONO_CLASS_GETTER(m_class_is_array_special_interface, gboolean, , MonoClass, is_array_special_interface)
 MONO_CLASS_GETTER(m_class_is_byreflike, gboolean, , MonoClass, is_byreflike)
+MONO_CLASS_GETTER(m_class_is_inlinearray, guint, , MonoClass, is_inlinearray)
```
Comment (vargaz) [inline src/mono/mono/metadata/class-getters.h:32]: The two types are switched.

File: src/mono/mono/metadata/class-init.c (hunk 1 above)
Comment (kotlarmilos) [reply src/mono/mono/metadata/class-init.c]: According to the test, static fields shouldn't be inlined, but you are right that it is not handled properly. Added a check  for empty structs in `mono_class_layout_fields`.

https://github.com/dotnet/runtime/blob/eefaa5a36af292a99be4e4421cb17e56e630f6cd/src/tests/Loader/classloader/InlineArray/InlineArrayInvalid.cs#L100-L116

File: src/mono/mono/metadata/object.c (hunk 3 above)
Comment (kotlarmilos) [reply src/mono/mono/metadata/object.c]: I agree. Would it make sense to do the following in case of inline array struct?
```c
bitmap [pos / BITMAP_EL_SIZE] |= ((gsize)1) << (pos % BITMAP_EL_SIZE);
*max_set = MAX (GINT_TO_UINT32(*max_set), pos);
```

If type is valuetype generic instance it falls through and invokes `compute_class_bitmap` recursively. By adding `type->type != MONO_TYPE_GENERICINST && type->type != MONO_TYPE_TYPEDBYREF && type->type != MONO_TYPE_VALUETYPE` it allows recursive call. 

If bitmap is calculated outside of switch statement, there is no need for type altering and generic instances condition.

File: src/mono/mono/metadata/class-init.c (hunk 2 above)
Comment (kotlarmilos) [reply src/mono/mono/metadata/class-init.c:798]: > I would prefer to pull this out into a callback for InlineArrayAttribute specifically.

Do you mean into `class_has_inlinearray_attribute` or to create a new callback for inlinearray attribute?

> I think this also needs some validation: the ctor has 1 argument, the argument really is a primitive, there are no named properties, stuff like that. otherwise this might be leaking decoded_attr fields.

Good point, will be added.

> Actually do you need to free `decoded_attr`?

Probably yes.

**Comment by kotlarmilos** [comment]
Inline array attribute tests have passed and the failure shouldn't be related.

**Comment by mangod9** [comment]
@VSadov too.

File: src/mono/mono/metadata/class-init.c [hunk 6]
```diff
@@ -746,65 +774,87 @@ mono_generic_class_setup_parent (MonoClass *klass, MonoClass *gtd)
...
 	 * MONO_RESOLUTION_SCOPE_ASSEMBLYREF and we'll need to check the
 	 * name.*/
-	return FALSE;
+	return attr->has_attr;
 }
 
-static gboolean
-class_has_wellknown_attribute (MonoClass *klass, const char *nspace, const char *name, gboolean in_corlib)
+static void
+has_wellknown_attribute_value_func (MonoImage *image, uint32_t method_token, uint32_t *cols, gpointer user_data)
```
Comment (lambdageek) [inline src/mono/mono/metadata/class-init.c:797]: I wonder if we should change the name to be more inlinearray specific.  it might work for other custom attributes, but it is pretty fragile: the class has to have 1 public ctor that users would use; the ctor must have 1 argument; the class probably must not have any public properties; the argument must be a primitive type.  In other cases using this method to extract the value would be a source of bugs.

File: src/mono/mono/metadata/class-init.c [hunk 7]
```diff
@@ -2217,6 +2271,16 @@ mono_class_layout_fields (MonoClass *klass, int base_instance_size, int packing_
 				}
 
 				size = mono_type_size (field->type, &align);
+				if (m_class_is_inlinearray (klass)) {
+					// limit the max size of array instance to 1MiB
+					const guint32 struct_max_size = 1024 * 1024;
+					size *= m_class_inlinearray_value (klass);
+					inlined_fields++;
+					if(size <= 0 || size > struct_max_size) {
+						mono_class_set_type_load_failure (klass, "Inline array struct size out of bounds.");
```
Comment (lambdageek) [inline src/mono/mono/metadata/class-init.c:2280]: Is this an error that we share with coreclr?  Should we mention the maximum size?

File: src/mono/mono/metadata/class-init.c [hunk 8]
```diff
@@ -2217,6 +2271,16 @@ mono_class_layout_fields (MonoClass *klass, int base_instance_size, int packing_
 				}
 
 				size = mono_type_size (field->type, &align);
+				if (m_class_is_inlinearray (klass)) {
+					// limit the max size of array instance to 1MiB
+					const guint32 struct_max_size = 1024 * 1024;
+					size *= m_class_inlinearray_value (klass);
+					inlined_fields++;
+					if(size <= 0 || size > struct_max_size) {
```
Comment (lambdageek) [inline src/mono/mono/metadata/class-init.c:2279]: nit: size is never < 0.  Size is a uint, and inlinearray_value is a uint.

File: src/mono/mono/metadata/class-private-definition.h [hunk 9]
```diff
@@ -102,6 +103,8 @@ struct _MonoClass {
 	guint16    *interface_offsets_packed;
 	guint8     *interface_bitmap;
 
+	guint inlinearray_value; /* System.Runtime.CompilerServices.InlineArrayAttribute lenght value */
```
[the remark this answers is not in the document]
Comment (lambdageek) [reply src/mono/mono/metadata/class-private-definition.h:106]: Actually... Should we use `gint32`?  The managed `InlineArrayAttribute` type uses `int` for the length.  so it might be negative.  and we will not notice negative values.  (well... it will error out saying they're out of bounds because a negative input will be a very large uint value.  But that's pretty fragile)

File: src/mono/mono/metadata/custom-attrs-internals.h [hunk 10]
```diff
@@ -35,7 +35,9 @@ typedef struct _MonoDecodeCustomAttr {
 MonoCustomAttrInfo*
 mono_custom_attrs_from_builders (MonoImage *alloc_img, MonoImage *image, MonoArray *cattrs);
 
-typedef gboolean (*MonoAssemblyMetadataCustomAttrIterFunc) (MonoImage *image, guint32 typeref_scope_token, const gchar* nspace, const gchar* name, guint32 method_token, gpointer user_data);
+typedef gboolean (*MonoAssemblyMetadataCustomAttrIterFunc) (MonoImage *image, guint32 typeref_scope_token, const gchar* nspace, const gchar* name, guint32 method_token, guint32 *cols, gpointer user_data);
+
+typedef void (*HasValueCallback) (MonoImage *image, uint32_t method_token, uint32_t *cols, gpointer user_data);
```
Comment (lambdageek) [inline src/mono/mono/metadata/custom-attrs-internals.h:40]: nit: `Mono` prefix on the typedef name.

**Review by lambdageek — APPROVED** [review]
LGTM.

The most important things to address:

1. we should make sure that `[InlineArray(-1)]` fails for the right reasons (because we correctly detect and disallow negative values) not because it wraps around to a huge unsigned value.
2. the callback that extracts the value shoudl be named more specifically for InlineArrayAttribute - I don't think it's a good idea to accidentally reuse it for another attribute that might be defined with more constructors, arguments or properties.

File: src/mono/mono/metadata/class-init.c [hunk 11]
```diff
@@ -291,6 +304,9 @@ mono_class_setup_fields (MonoClass *klass)
 		instance_size = MONO_ABI_SIZEOF (MonoObject);
 	}
 
+	if (m_class_is_inlinearray (klass) && m_class_inlinearray_value (klass) <= 0)
+		mono_class_set_type_load_failure (klass, "Inline array lenght must be greater than 0.");
```
Comment (VSadov) [inline src/mono/mono/metadata/class-init.c:308]:
```suggestion
		mono_class_set_type_load_failure (klass, "Inline array length must be greater than 0.");
```

File: src/mono/mono/metadata/class-init.c (hunk 7 above)
Comment (VSadov) [reply src/mono/mono/metadata/class-init.c:2280]: > Should we mention the maximum size?

I am not sure the limit is very informative to the user.

The size is computed from element size and will be impacted by aligning and padding. The limit was rather arbitrarily chosen as there would be some limit anyways. 
1 MiB seems to be enough for most uses. It can be increased if there is a need for more, but likely this error will be seen only in erroneous cases.  

We basically want to say "Hey, the array is abnormally large. Is everything alright with the code?"

File: src/mono/mono/metadata/object.c [hunk 12]
```diff
@@ -931,6 +932,15 @@ compute_class_bitmap (MonoClass *klass, gsize *bitmap, int size, int offset, int
 				g_error ("compute_class_bitmap: Invalid type %x for field %s:%s\n", type->type, mono_type_get_full_name (m_field_get_parent (field)), field->name);
 				break;
 			}
+
+			// If a struct has inline array attribute, consider it as an array
```
Comment (VSadov) [inline src/mono/mono/metadata/object.c:936]: I am not familiar with how this GC/pointer maps work in Mono. The rest of the change looks good to me.

**Review by VSadov — APPROVED** [review]

File: src/mono/mono/metadata/class-private-definition.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (kotlarmilos) [reply src/mono/mono/metadata/class-private-definition.h:106]: Right, it should be gint32 to detect negative values as well.

File: src/libraries/System.Runtime.InteropServices.JavaScript/gen/JSImportGenerator/JSManagedTypeInfo.cs [hunk 13]
```diff
@@ -163,6 +163,10 @@ public static JSTypeInfo CreateJSTypeInfoForTypeSymbol(ITypeSymbol type)
                         return new JSInvalidTypeInfo();
                     }
                     return new JSFunctionTypeInfo(false, signatureTypes);
+
+                // struct with inline array attribute
+                case { TypeKind: TypeKind.Struct } structType when structType.GetAttributes().Any(attr => attr.AttributeClass?.ToDisplayString() == "System.Runtime.CompilerServices.InlineArrayAttribute"):
```
Comment (marek-safar) [inline src/libraries/System.Runtime.InteropServices.JavaScript/gen/JSImportGenerator/JSManagedTypeInfo.cs:168]: Why is it needed if it's same as the fallback?

File: src/mono/mono/metadata/class-init.c (hunk 6 above)
Comment (kotlarmilos) [reply src/mono/mono/metadata/class-init.c:797]: Renamed to `has_inline_array_attribute_value_func`.

File: src/mono/mono/metadata/class-init.c (hunk 7 above)
Comment (kotlarmilos) [reply src/mono/mono/metadata/class-init.c:2280]: It might be valuable for users to provide more information on the error, either maximum size or indication that the struct is abnormally large. Would it be wrong to mention maximum size as it could be impacted by aligning and padding?

File: src/mono/mono/metadata/object.c (hunk 12 above)
Comment (kotlarmilos) [reply src/mono/mono/metadata/object.c:936]: A struct with the inline array attribute is handled in the same way as an array.

File: src/libraries/System.Runtime.InteropServices.JavaScript/gen/JSImportGenerator/JSManagedTypeInfo.cs (hunk 13 above)
Comment (kotlarmilos) [reply src/libraries/System.Runtime.InteropServices.JavaScript/gen/JSImportGenerator/JSManagedTypeInfo.cs:168]: Good point. Currently, the JS Interop generator does not support the marshalling of structs. This was done to allow for potential future extensibility, to disallow marshalling of structs with the ` InlineArrayAttribute` in case structs were to be allowed for marshalling in the future.

However, if it is not necessary, we can leave a comment as a reminder.

**Comment by kotlarmilos** [comment]
Failures shouldn't be related.
